Implement Citus compatibility in DCS (#7)
CyberDem0n/patroni: docs/citus.rst

.. _citus:

Citus support
=============

> Patroni makes it extremely simple deploying `Multi-Node Citus`__ clusters.

Reviewer: ```suggestion
Patroni makes it extremely simple to deploy `Multi-Node Citus`__ clusters.
```

__ https://docs.citusdata.com/en/stable/installation/multi_node.html

TL;DR
-----

There are only a few simple rules you need to follow:

> 1. The Citus extension must be available on all nodes.
> 2. The cluster name (``scope``) must be the same for all Citus nodes!
> 3. The superuser credentials must be the same on coordinator and all worker
>    nodes, and ``pg_hba.conf`` should allow superuser access between all nodes.

Reviewer: ```suggestion
1. Citus extension must be available on all nodes.
2. Cluster name (``scope``) must be the same for all Citus nodes!
3. Superuser credentials must be the same on coordinator and all worker
   nodes, and ``pg_hba.conf`` should allow superuser access between all nodes.
```

4. :ref:`REST API <restapi_settings>` access should be allowed from worker

>    nodes to the coordinator. E.g., credentials should be the same and if
>    configured client certificates from worker nodes must be accepted by the

Reviewer: ```suggestion
   nodes to the coordinator. E.g., credentials should be the same and if
   configured, client certificates from worker nodes must be accepted by the
```

   coordinator.
5. Add the following section to the ``patroni.yaml``:

.. code:: YAML

        citus:
          group: X  # 0 for coordinator and 1, 2, 3, etc for workers
          database: citus  # must be the same on all nodes

After that you just need to start Patroni and it will handle the rest:

> 1. The ``citus`` extension will be automatically added to the ``shared_preload_libraries``.

Reviewer: ```suggestion
1. ``citus`` extension will be automatically added to ``shared_preload_libraries``.
```

2. If ``max_prepared_transactions`` isn't explicitly set in the global
   :ref:`dynamic configuration <dynamic_configuration>` Patroni will
   automatically set it to ``2*max_connections``.
3. The ``citus.database`` will be automatically created followed by ``CREATE EXTENSION citus``.
4. Current superuser :ref:`credentials <postgresql_settings>` will be added to the ``pg_dist_authinfo``
   table to allow cross-node communication. Don't forget to update them if
   later you decide to change superuser username/password/sslcert/sslkey!
5. The coordinator primary node will automatically discover worker primary
   nodes and add them to the ``pg_dist_node`` table using the
   ``citus_add_node()`` function.

> 6. Patroni will also maintain ``pg_dist_node`` in case if failover/switchover

Reviewer: ```suggestion
6. Patroni will also maintain ``pg_dist_node`` in case failover/switchover
```

    postgres@coord1:~$ patronictl list demo --group 0
    + Citus cluster: demo (group: 0, 7179854923829112860) -----------+
    | Member | Host        | Role         | State   | TL | Lag in MB |
    +--------+-------------+--------------+---------+----+-----------+
    | coord1 | 172.27.0.10 | Replica      | running |  1 |         0 |
    | coord2 | 172.27.0.6  | Sync Standby | running |  1 |         0 |
    | coord3 | 172.27.0.4  | Leader       | running |  1 |           |
    +--------+-------------+--------------+---------+----+-----------+

    postgres@coord1:~$ patronictl list demo --group 1
    + Citus cluster: demo (group: 1, 7179854923881963547) -----------+
    | Member  | Host       | Role         | State   | TL | Lag in MB |
    +---------+------------+--------------+---------+----+-----------+
    | work1-1 | 172.27.0.8 | Sync Standby | running |  1 |         0 |
    | work1-2 | 172.27.0.2 | Leader       | running |  1 |           |
    +---------+------------+--------------+---------+----+-----------+

Citus worker switchover
-----------------------

When a switchover is orchestrated for a Citus worker node, Citus offers the

> opportunity to make the switchover close to transparent to the application.

Reviewer: ```suggestion
opportunity to make the switchover close to transparent for an application.
```

    # The future worker primary notifies the coordinator that it acquired the leader lock in DCS and about to run "pg_ctl promote".
    2022-12-22 07:02:40,085 DEBUG: query("SELECT pg_catalog.citus_update_node(3, '172.27.0.5', 5432)")

    # The new worker primary just finished promote and notifies coordinator that it is ready to accept read-write traffic.
    2022-12-22 07:02:41,485 DEBUG: query("COMMIT")
    # From this moment the application traffic on the coordinator to the worker group 2 is unblocked.

Peek into DCS
-------------

The Citus cluster (coordinator and workers) are stored in DCS as a fleet of
Patroni clusters logically grouped together::

    /service/batman/              # scope=batman
    /service/batman/0/            # citus.group=0, coordinator
    /service/batman/0/initialize
    /service/batman/0/leader
    /service/batman/0/members/
    /service/batman/0/members/m1
    /service/batman/0/members/m2
    /service/batman/1/            # citus.group=1, worker
    /service/batman/1/initialize
    /service/batman/1/leader
    /service/batman/1/members/
    /service/batman/1/members/m3
    /service/batman/1/members/m4
    ...

> Such an approach was choosen because for most DCS it becomes possible to fetch

Reviewer: ```suggestion
Such an approach was chosen because for most DCS it becomes possible to fetch
```

            cluster-name: citusdemo
          name: citusdemo-2-0
          namespace: default
        spec:
          containers:
          - env:
            - name: PATRONI_SCOPE
              value: citusdemo
            - name: PATRONI_NAME
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.name
            - name: PATRONI_KUBERNETES_POD_IP
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: status.podIP
            - name: PATRONI_KUBERNETES_NAMESPACE
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
            - name: PATRONI_KUBERNETES_LABELS
              value: '{application: patroni}'
            - name: PATRONI_CITUS_DATABASE
              value: citus
            - name: PATRONI_CITUS_GROUP
              value: "2"

As you may noticed, both examples have ``citus-group`` label set. This label
allows Patroni to identify object as belonging to a certain Citus group. In
addition to that, there is also ``PATRONI_CITUS_GROUP`` environment variable,

> which has the same value as the ``citus-group`` label. When Patroni will create
> new Kubernetes objects ConfigMaps or Endpoints, it will automatically put the

Reviewer: ```suggestion
which has the same value as the ``citus-group`` label. When Patroni creates
new Kubernetes objects ConfigMaps or Endpoints, it automatically puts the
```
